rules/terraform_required_version.go

```diff
@@ -78,13 +78,32 @@ func (r *TerraformRequiredVersionRule) Check(runner tflint.Runner) error {
 		exists = exists || ok
 	}
 
-	if !exists {
-		return runner.EmitIssue(
-			r,
-			`terraform "required_version" attribute is required`,
-			hcl.Range{},
-		)
+	if exists {
+		return nil
+	}
+
+	var missingRange hcl.Range
+
+	if len(body.Blocks) > 0 {
+		missingRange = body.Blocks[0].DefRange
+	} else {
+		// If there are no "terraform" blocks, create a hcl.Range for a file
+
+		// Grab a single filename from keys of files map
+		var filename string
+		for k := range files {
+			filename = k
+			break
+		}
+
+		missingRange = hcl.Range{
+			Filename: filename,
+		}
```

Comment:
Well, in my case at least, completely skip this logic wouldn't solve the issue I ran into #177. I had several local modules and none of them had a `"terraform"` block. When I ran tflint for the first time I got a bunch of

```
Warning: terraform "required_version" attribute is required (terraform_required_version)

  on  line 0:
   (source code not available)
```

And was very confused about where these were coming from.

I agree that a random file is not ideal. What about pointing to the directory?

Reply:
In theory, `Filename` should be a filename. If you can show a directory contributes useful output and doesn't break things that sounds fine.

Reply:
Alright, I've updated the else logic. In the case where files are missing a `"terraform"` block, it now references either the single file in the module or it points to the directory if there are multiple files. The test cases "no terraform block" and "multiple files no terraform blocks" cover these scenarios. Without the else logic, both of those cases would report the confusing range: `  on  line 0:`.

In my local example that lead to the creation of the issue, the output now looks like:

```
❯ tflint --recursive --only terraform_required_version
2 issue(s) found:

Warning: terraform "required_version" attribute is required (terraform_required_version)

  on modules/deployment/ line 0:
   (source code not available)

Reference: https://github.com/terraform-linters/tflint-ruleset-terraform/blob/v0.7.0/docs/rules/terraform_required_version.md

Warning: terraform "required_version" attribute is required (terraform_required_version)

  on modules/dns/ line 0:
   (source code not available)

Reference: https://github.com/terraform-linters/tflint-ruleset-terraform/blob/v0.7.0/docs/rules/terraform_required_version.md

Warning: terraform "required_version" attribute is required (terraform_required_version)

  on modules/email/ line 0:
   (source code not available)

Reference: https://github.com/terraform-linters/tflint-ruleset-terraform/blob/v0.7.0/docs/rules/terraform_required_version.md
```

Reply:
As mentioned in https://github.com/terraform-linters/tflint/issues/1790#issuecomment-2135619986, I favor a random file over passing directory path as `Filename`.
I agree that this is not ideal, but I think it's a reasonable choice considering the integration with other tools.

However, we should always attach the same range so that it does not change each time you run it.

Reply:
@wata727 I'm happy to make that change. My thinking on that would be to point to `main.tf` if it exists in the module (or is there something else that'd be preferred? or maybe an ordered list of files to check for?) and then fallback to selecting the first file alphabetically otherwise. Based on your other comment, should these emitted ranges also include a `Pos` for the `Start` and `End`? When the `"terraform"` block is completely missing, maybe it makes sense to have the start and end point to the start of the file?

I welcome your input

Reply:
> My thinking on that would be to point to main.tf if it exists in the module and then fallback to selecting the first file alphabetically otherwise.

I support this idea. If possible, I would also like to hear @bendrucker's opinion.

> should these emitted ranges also include a Pos for the Start and End? When the "terraform" block is completely missing, maybe it makes sense to have the start and end point to the start of the file?

Yeah, I think it's helpful to attach to the "terraform" block range. If there is no "terraform" block, it is probably better to use [`hcl.InitialPos`](https://github.com/hashicorp/hcl/blob/v2.20.1/pos.go#L38) as `Pos`. You can also use annotations such as `tflint-ignore` by pointing to the first line.

Reply:
Maybe `main.tf` if it's there otherwise preserve the current behavior. Encouraging a random file, which will need to be alphabetically first to be deterministic, is unpredictable and awkward.

Reply:
An arbitrary file may be awkward, sure, but IMHO certainly better than providing no context for the source of the issue at all.

In my case pointing to `main.tf` would've resolved the issue in #177, so I'm happy to do that, but want to share my 2 cents.

Reply:
Last idea is to report on a `main.tf` file, whether or not it's there. The source code won't be there but you'll get a range printed with `(source unavailable)` or similar.

> certainly better than providing no context

The other side of this coin is that static analysis tools have a duty to not provide wrong advice. False detection positives will happen but the remediation advice should be clear/specific. Picking an arbitrary file and telling the user "put this config there" seems to risk giving bad advice (more specific than the context allows).

Reply:
> Picking an arbitrary file and telling the user "put this config there" seems to risk giving bad advice (more specific than the context allows).

Agreed with this concern. My proposed policy doesn't allow for the issue to be "create a new file and write to it."

From the perspective of avoiding false positives, maybe there is a way to not report an issue if the "terraform" block is missing. However, I'm not sure if this is the behavior that users want.

On the other hand, any potential misleading behavior may be mitigated by other rules such as `terraform_standard_module_structure`.

> Last idea is to report on a main.tf file, whether or not it's there. The source code won't be there but you'll get a range printed with (source unavailable) or similar.

I've thought about this idea too, but ultimately it brings up issues that aren't well-handled in LSP or Problem Matchers, so I'm not positive about it.

Reply:
As an aside, instead of `main.tf`, maybe you could follow the official guidelines and recommend `terraform.tf` instead.
https://developer.hashicorp.com/terraform/language/style#file-names

Reply:
Oh nice! Pretty recent, let's do that:

https://github.com/hashicorp/terraform/pull/34876